// @liveblocks/react-ui/dist/components/InboxNotificationList.js
"use client";
'use strict';

var jsxRuntime = require('react/jsx-runtime');
var react = require('react');
var classNames = require('../utils/class-names.js');
var useVisible = require('../utils/use-visible.js');


function ReachEndMarker({
  enabled,
  onReachEnd
}) {
  const markerRef = react.useRef(null);
  useVisible.useVisibleCallback(markerRef, onReachEnd, {
    enabled
  });
  return /* @__PURE__ */ jsxRuntime.jsx("div", {
    ref: markerRef,
    style: { height: 0 }
  });
}
const InboxNotificationList = react.forwardRef(({ onReachEnd, children, className, ...props }, forwardedRef) => {
  return /* @__PURE__ */ jsxRuntime.jsxs("ol", {
    className: classNames.classNames("lb-root lb-inbox-notification-list", className),
    ...props,
    ref: forwardedRef,
    children: [
      react.Children.map(children, (child, index) => /* @__PURE__ */ jsxRuntime.jsx("li", {
        className: "lb-inbox-notification-list-item",
        children: child
      }, index)),
      onReachEnd && /* @__PURE__ */ jsxRuntime.jsx(ReachEndMarker, {
        onReachEnd,
        enabled: react.Children.count(children) > 0
      })
    ]
  });
});

exports.InboxNotificationList = InboxNotificationList;
